### ISSUE TITLE
bde_ApplyLevel0Update drops BDE function

### ISSUE THREAD

**Issue by strk** [linz/linz-bde-uploader#146]
The `bde_ApplyLevel0Update` function does a `DROP TABLE xx CASCADE` which cascades to dropping the `bde_get_app_specific(crs_appellation, char)`.

The author of that function (@palmerj?) had dependent objects in mind as there is code to "get dependent objects" and later code to "restore dependent objects". Only the "get dependent objects" function is just an empty stub: 

```
CREATE OR REPLACE FUNCTION _bde_GetDependentObjectSql(
    p_upload INTEGER,
    p_base regclass
)
RETURNS
    text[]
AS  
$body$ 
BEGIN
    RETURN ARRAY[]::text[];
END 
$body$
LANGUAGE plpgsql IMMUTABLE; -- pointless empty function ?
-- TODO: does this function need to be implemented ?
--       Would become STABLE then...
```

I guess the comment was mine when I went through all functions to mark them STABLE/VOLATILE.

Now I can see the function was not pointless, just unimplemented.

This is an urgent matter as it's breaking current LDS uploads due to MaintainSimplifiedParcelLayer not finding the function (because it's dropped here). See https://github.com/linz/linz-lds-bde-schema/issues/65

**Comment by strk** [comment]
I'd really like to simplify the code and avoid `DROP TABLE` completely.

Jeremy could you clarify why `DROP TABLE old` and `RENAME TABLE new TO old` trick is there instead of a `TRUNCATE old' and 'INSERT INTO old SELECT * from new` ?

**Comment by palmerj** [comment]
> I'd really like to simplify the code and avoid DROP TABLE completely.

Ok I would do that

> Jeremy could you clarify why DROP TABLE old and RENAME TABLE new TO old trick is there instead of a TRUNCATE old' and 'INSERT INTO old SELECT * from new ?

I believe it was to ensure that no WAL logging was done on loading the data and it was decided not to use temp tables due to some other reason I've now forgotten (maybe debugging the data state after each step?). But as you say TRUNCATE and then INSERT INTO if done in the same transaction should give the same result

**Comment by strk** [comment]
Well a temp table is actually used ("new" table).
Doing TRUNCATE and INSERT INTO would take more time unless we really drop the temporary table and do the insert directly into the destination

**Comment by strk** [comment]
I'm getting this in production testing:

> function bde.bde_get_app_specific(bde.crs_appellation,character) depends on type bde.crs_appellation

When removing the CASCADE part. Of course we can re-add the CASCADE, but we're clearly dropping a function here, if we do that.

**Comment by strk** [comment]
I'll try to implement the `_bde_GetDependentObjectSql` stub and see if it helps fixing this problem, which now became urgent as it prevents using `table_version` 1.6.0 successfully unless we revert to the destructive `CASCADE` drop.

**Comment by strk** [comment]
It looks like I already implemented _bde_GetDependentObjectSql in f0b32897340aa5c73c88cc7b63ceebb6e2cea707 which supposedly fixed this ticket, but lacking an automated test.

Now, this means that the function being DROPed by CASCADE will be created again, so we should be fine with this for now.
